Implement License verification method

smakalias opened the pull request:
## Summary of the changes / Why this improves CrateDB

Implement the the license verification method.

## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [ ] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed

smakalias commented:
WIP: Early push for visibility and perhaps initial comments.

Next steps:
- Asymmetric keys generation
- Testing

andreidan commented on licensing/src/main/java/io/crate/license/LicenseService.java:
```diff
@@ -129,6 +122,81 @@ public void onFailure(Exception e) {
+                return false;
+            }
+            if (decodedLicense.type() == LicenseType.ENTERPRISE) {
+                return verifySignature(licenseInfo);
```
To me it feels a bit redundant to encrypt the content and have a signature as well but maybe I'm missing something. The content we would encrypt with the private key could be unencrypted with the public key we ship and I think that's the only thing that should differ from the self generated license processing. 

Maybe I am missing something though. @mfussenegger what do you think?

mfussenegger replied:
It's a bit difficult to see the full license verification flow with just the bits and pieces in the diffs.

I guess it would make sense to have some high level javadocs that illustrates the flow?

smakalias replied:
added the high level javadocs.

It is true that in the presence of the signature (enterprise licenses), encryption feels a bit redundant, however I decided to keep it as 
a) it's not that expensive (symmetric)
b) it makes handling the two types of licenses more consistent -> cleaner code

but I'm open to suggestions :)

mfussenegger commented on licensing/src/main/java/io/crate/license/Cryptos.java:
```diff
@@ -30,17 +32,49 @@
+    /**
+     * Read public key file content
+     */
+    static PublicKey readPublicKey(byte[] fileContents) {
```
`read` is imho miss-leading, it's not reading anything as the file contents are already provided?

smakalias replied:
done

mfussenegger commented on licensing/src/main/java/io/crate/license/Cryptos.java:
```diff
@@ -30,17 +32,49 @@
+    /**
+     * Read private key file content
+     */
+    static PrivateKey readPrivateKey(byte[] fileContents) {
```
Same here.. no `read` involved here

mfussenegger commented on licensing/src/main/java/io/crate/license/Cryptos.java:
```diff
@@ -62,4 +96,11 @@ private Cryptos() {
     }
+
+    static byte[] digest(byte[] inputBytes) {
+        MessageDigest sha256 = MessageDigests.sha256();
```
We don't need the thread-local facility of `MessageDigests` afaik - so we could go with plain `MessageDigest.getInstance(digest)`. I wonder if we should expose the used hash function to the API user? E.g. `sha256Digest(..)` ?

mfussenegger commented on licensing/src/main/java/io/crate/license/DecryptedLicenseData.java:
```diff
@@ -31,18 +32,28 @@
 
     private final long expirationDateInMs;
     private final String issuedTo;
+    private Optional<String> signature;
```
We're usually going with `Nullable` annotation and nullable objects instead of using `Optional`

mfussenegger commented on licensing/src/main/java/io/crate/license/DecryptedLicenseData.java:
```diff
@@ -53,23 +64,68 @@ String issuedTo() {
+        return (signature.isPresent() ? signature.get() : "");
+    }
+
+    void setSignature(String signature) {
```
Why do we need a setter for this?

smakalias replied:
It is used by the service [to bundle the newly created signature to the license](https://github.com/crate/crate/pull/7758/files#diff-4b272f5578419e39b7c0a1db3b7ca030R195).

mfussenegger commented on licensing/src/main/java/io/crate/license/DecryptedLicenseData.java:
```diff
@@ -53,23 +64,68 @@ String issuedTo() {
     }
 
+    String signature() {
+        return (signature.isPresent() ? signature.get() : "");
```
`Optional` would have a `orElse` method. (But see my other comment about usually not using it)

mfussenegger commented on licensing/src/main/java/io/crate/license/DecryptedLicenseData.java:
```diff
@@ -53,23 +64,68 @@ String issuedTo() {
-                .field(ISSUED_TO, issuedTo)
-                .endObject();
+                .field(ISSUED_TO, issuedTo);
+            if (includeSignature) {
```
Shouldn't we always include it if it is present? Why is it okay to throw away the signature?

smakalias replied:
At the time of the verification, where the signature is present at the license, we need to generate a new signature based on the  license content (exluding the signature itself) so that we test signature validity. In that case we need to skip the signature.

It was the reason that I insisted of having another class i.e. LicenseInfo containing only the actual license information (ie expirationDate, issuedTo). I think this wouls be much cleaner that way.

mfussenegger commented on licensing/src/main/java/io/crate/license/LicenseKey.java:
```diff
@@ -35,13 +35,44 @@
+            this.value = value;
+        }
+
+        static LicenseType of(Integer value) {
```
`Integer -> int` primitive type?

mfussenegger commented on licensing/src/main/java/io/crate/license/LicenseKey.java:
```diff
@@ -35,13 +35,44 @@
+        static LicenseType of(Integer value) {
+            LicenseType type = types.get(value);
+            if (type == null) {
+                throw new InvalidLicenseException("Invalid License Type");
```
Maybe include the given `value` in the error message?

mfussenegger commented on licensing/src/main/java/io/crate/license/LicenseKey.java:
```diff
@@ -35,13 +35,44 @@
+        }
+
+        static LicenseType of(Integer value) {
+            LicenseType type = types.get(value);
```
I wonder if we should do a simple `switch/case` here instead of going through the map? Or simply use the Enum ordinals.

smakalias replied:
Agree. Map is an overkill. Changed this to use a simple switch statement, I'm not that fond of using Enum ordinals, it can lead to issues (eg if the order changes in the future).

mfussenegger commented on licensing/src/main/java/io/crate/license/LicenseService.java:
```diff
@@ -129,6 +122,81 @@ public void onFailure(Exception e) {
+    private static boolean verifySignature(final DecryptedLicenseData licenseData, byte[] publicKeyBytes) {
+        try {
+            // init signature with public key
+            Signature rsa = Signature.getInstance("SHA512withRSA");
```
Would it be possible to split the crypto code more from the licensing specific code?

We're creating loose contracts here in that the used algorithms have to match. E.g. `Cryptos` has a `RSA` constant and here we're defining `SHA512WithRSA` but there is no gurantee that the `readPublicKey` call returns a public key that "fits"

smakalias replied:
done

smakalias commented:
pushed some fixes and added some baseline tests to validate correctness
Next:
- Address/review the comments
- Add more tests

smakalias commented:
@mfussenegger @andreidan 

Addressed almost all*  comments, brought it up to date with feature branch and added/cleanup tests
*Only thing missing is the "high level javadocs". Will add it asap.

smakalias commented:
@mfussenegger @andreidan 
please have another look

smakalias commented on licensing/src/main/java/io/crate/license/CryptoUtils.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    /**
+     * Get a {@link PrivateKey} from the encrypted key content
```
no, comments were left from previous version of the methods - removed javadocs completely as - I agree - they don't add much anyway

smakalias commented on licensing/src/main/java/io/crate/license/CryptoUtils.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    /**
+     * Get a {@link PrivateKey} from the encrypted key content
```
no, comments were left from previous version of the methods - removed javadocs completely as - I agree - they don't add much anyway

smakalias commented on licensing/src/main/java/io/crate/license/CryptoUtils.java:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+
+    static void writeKeysToFiles(final KeyPair keyPair,
```
Moved it to tests. Actually it canoe removed (together with the test itself) but decided to leave it for now as a quick way of generating the rsa keypair for tests. I plan to remove it on a later stage.

smakalias commented on licensing/src/main/java/io/crate/license/EnterpriseLicense.java:
```diff
@@ -0,0 +1,64 @@
+    private EnterpriseLicense() {
+    }
+
+    static LicenseKey createLicenseKey(int version, DecryptedLicenseData decryptedLicenseData,
```
done

smakalias commented on licensing/src/main/java/io/crate/license/CryptoUtils.java:
```diff
@@ -0,0 +1,158 @@
+            data);
+    }
+
+    static byte[] encryptRsaUsingPrivateKey(byte[] data, byte[] privateKeyBytes) {
```
done

smakalias commented on licensing/src/main/java/io/crate/license/LicenseKey.java:
```diff
@@ -41,12 +41,37 @@
+                    return LicenseType.ENTERPRISE;
+                default:
+            }
+            throw new InvalidLicenseException("Invalid License Type of value: " + value);
```
there was no complain, however I did move the throw stmt as it makes more sense to be there. thanks

smakalias commented on licensing/src/test/java/io/crate/license/DecryptedLicenseDataTest.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @Test
+    public void testFormatAndFromFormat() throws IOException {
```
My intention was to test the serialisation/de-serialisation in that case: `formatLicenseData()` and `fromFormattedLicenseData(byte[])`

I am testing the pair of methods together so I don't introduce byte[] intermediate values.
I will change the test name, do you think I should also split this into 2 tests?

mfussenegger commented on licensing/src/test/java/io/crate/license/CryptoUtilsTest.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    @Test
+    public void testWriteKeysToFilesDoesProduceFiles() throws IOException {
```
Could we remove this ? I don't see the point of testing a method that's only in the test class

smakalias replied:
done
